pkg/volume/scaleio/sio_util.go

```diff
@@ -0,0 +1,249 @@
+		storagePool:      "storagePool",
+		storageMode:      "storageMode",
+		sdcRootPath:      "sdcRootPath",
+		volumeName:       "volumeName",
+		volSpecName:      "volSpecName",
+		fsType:           "fsType",
+		readOnly:         "readOnly",
+		username:         "username",
+		password:         "password",
+	}
+	nsSep       = "%"
+	sdcRootPath = "/opt/emc/scaleio/sdc/bin"
+
+	secretNotFoundErr       = errors.New("secret not found")
+	configMapNotFoundErr    = errors.New("configMap not found")
+	gatewayNotProvidedErr   = errors.New("gateway not provided")
+	secretRefNotProvidedErr = errors.New("secret ref not provided")
+	systemNotProvidedErr    = errors.New("secret not provided")
+)
+
+// mapScaleIOVolumeSource maps attributes from a ScaleIOVolumeSource to config
+func mapVolumeSource(config map[string]string, source *api.ScaleIOVolumeSource) {
+	config[confKey.gateway] = source.Gateway
+	config[confKey.secretRef] = func() string {
+		if source.SecretRef != nil {
+			return string(source.SecretRef.Name)
+		}
+		return ""
+	}()
+	config[confKey.system] = source.System
+	config[confKey.volumeName] = source.VolumeName
+	config[confKey.sslEnabled] = strconv.FormatBool(source.SSLEnabled)
+	config[confKey.protectionDomain] = source.ProtectionDomain
+	config[confKey.storagePool] = source.StoragePool
+	config[confKey.storageMode] = source.StorageMode
+	config[confKey.fsType] = source.FSType
+	config[confKey.readOnly] = strconv.FormatBool(source.ReadOnly)
+
+	//optionals
+	applyConfigDefaults(config)
+}
+
+func validateConfigs(config map[string]string) error {
+	if config[confKey.gateway] == "" {
+		return gatewayNotProvidedErr
+	}
+	if config[confKey.secretRef] == "" {
+		return secretRefNotProvidedErr
+	}
+	if config[confKey.system] == "" {
+		return systemNotProvidedErr
+	}
+
+	return nil
+}
+
+//TODO remove, default check is done by k8s type checks
+func applyConfigDefaults(config map[string]string) {
+	b, err := strconv.ParseBool(config[confKey.sslEnabled])
+	if err != nil {
+		glog.Warning(log("failed to parse param sslEnabled, setting it to false"))
+		b = false
+	}
+	config[confKey.sslEnabled] = strconv.FormatBool(b)
+	config[confKey.protectionDomain] = defaultString(config[confKey.protectionDomain], "default")
+	config[confKey.storagePool] = defaultString(config[confKey.storagePool], "default")
+	config[confKey.storageMode] = defaultString(config[confKey.storageMode], "ThinProvisioned")
+	config[confKey.fsType] = defaultString(config[confKey.fsType], "xfs")
+	b, err = strconv.ParseBool(config[confKey.readOnly])
+	if err != nil {
+		glog.Warning(log("failed to parse param readOnly, setting it to true"))
+		b = true
+	}
+	config[confKey.readOnly] = strconv.FormatBool(b)
+}
+
+func defaultString(val, defVal string) string {
+	if val == "" {
+		return defVal
+	}
+	return val
+}
+
+// saveConfigMap creates/stores a configMap from the config data
+func saveConfigMap(plug *sioPlugin, namespace string, configName string, data map[string]string) error {
```

Comment:
@rootfs this function is called couple of times in pkg/volume/scaleio/sio_volume.go